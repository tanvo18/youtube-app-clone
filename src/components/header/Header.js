import React, { Component } from 'react';
import logoImg from '../../assets/images/youtube.png';
import Logo from '../common/logo/Logo';
import Input from '../common/input/Input';
import Button from '../common/button/Button';
import HeaderStyled from './HeaderStyled';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { calRem } from '../../themes/mixins';

class Header extends Component {
  constructor(props) {
    super(props);
    this.inputRef = null;
  }

  /**
   * Set input ref
   */
  setInputRef = (element) => {
    this.inputRef = element;
  }

  /**
   * Handle key press on search bar
   */
  handleKeyDown = (event) => {
    if (event.keyCode === 13) {
      // Redirect to search
      this.props.history.push('/search');

      // Search video
      this.props.searchVideo(this.inputRef.value);
    }
  }

  handleAuthClick = () => {
    this.props.loginRequest();
  }

  handleSignoutClick = () => {
    this.props.logoutRequest();
  }

  render() {
    const { login, user } = this.props;

    return (
      <HeaderStyled className="container-fluid">
        <Logo
          link="/"
          logoClass="logo"
          logoImg={logoImg}
          logoWidth='80'
          logoHeight='64'
        />
        <div className="search-wrap">
          <Input
            inputType="text"
            placeHolder="Search"
            inputRef={this.setInputRef}
            borderRadius="0"
            inputWidth="100%"
            inputHeight={`${calRem(38)}`}
            handleKeyDown={this.handleKeyDown}
          />

          <div className="input-group-append">
            <Link
              to={'/search'}
            >
              <Button
                btnClass="btn"
                text={<i className="fa fa-search"></i>}
                borderRadius="0"
                borderColor="#ccd3d9"
                hoverBgColor="#d9d9d9"
                textColor='gray'
                handleClick={() => this.props.searchVideo(this.inputRef.value)}
              />
            </Link>
          </div>
        </div>

        <div className="login-btn-wrap">
          <div>
            {login.isSignedIn
              ? <div>
                <div className="user-info dropdown">
                  <button className="btn-avatar dropdown-toggle" data-toggle="dropdown">
                    <img src={user.user.picture} alt="" />
                  </button>

                  <div className="dropdown-menu">
                    <Link
                      to={'/playlist'}
                      className="dropdown-item"
                    >
                      Playlist
                    </Link>
                    <Link
                      to={'/upload-video'}
                      className="dropdown-item"
                    >
                      Upload video
                    </Link>
                    <Link
                      to={'/'}
                      className="dropdown-item logout-link"
                      onClick={this.handleSignoutClick}
                    >
                      Logout
                    </Link>
                  </div>
                </div>

              </div>
              : <Button
                btnClass="btn btn-login"
                text="SIGN IN"
                btnBackground="transparent"
                textColor="#0194E1"
                handleClick={this.handleAuthClick}
              />
            }
          </div>
        </div>
      </HeaderStyled>
    );
  }
};

Header.defaultProps = {
  login: {},
  user: {},
  searchVideo: () => { },
  loginRequest: () => { },
  logoutRequest: () => { },
};

Header.propTypes = {
  login: PropTypes.object,
  user: PropTypes.object,
  searchVideo: PropTypes.func,
  loginRequest: PropTypes.func,
  logoutRequest: PropTypes.func,
};

export default Header;
