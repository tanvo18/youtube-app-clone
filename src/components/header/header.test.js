import React from 'react';
import Header from './Header';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';

const props = {
  user: {
    'user': {
      'id': '103253251803252352082',
      'email': 'test@gmail.com',
      'verified_email': true,
      'name': 'Tan Vo',
      'given_name': 'Tan',
      'family_name': 'Vo',
      'link': 'https://plus.google.com/103253251803252352082',
      'picture': 'https://lh4.googleusercontent.com/-RZTNibkjm88/AAAAAAAAAAI/AAAAAAAAAJc/orvzalcw0uY/photo.jpg',
      'gender': 'male',
      'locale': 'en'
    },
    'isFetching': false,
    'error': null,
  },
};

describe('test header', () => {
  it('test function searchVideo works', () => {
    const login = {
      isSignedIn: true
    };
    const searchVideo = jest.fn();
    const wrapper = mount(<Router>
      <Header login={login} searchVideo={searchVideo} {...props} />
    </Router>);
    wrapper.find('Button').simulate('click');
    expect(searchVideo).toHaveBeenCalled();
  });

  it('test function loginRequest works', () => {
    const login = {
      isSignedIn: false
    };
    const loginRequest = jest.fn();
    const wrapper = mount(<Router>
      <Header login={login} loginRequest={loginRequest} {...props} />
    </Router>);
    const btnLogin = wrapper.find('.btn-login').at(1);
    btnLogin.simulate('click');
    expect(loginRequest).toHaveBeenCalled();
  });

  it('test function logoutRequest works', () => {
    const login = {
      isSignedIn: true
    };
    const logoutRequest = jest.fn();
    const wrapper = mount(<Router>
      <Header
        login={login}
        logoutRequest={logoutRequest}
        {...props} />
    </Router>);
    const logoutLink = wrapper.find('.logout-link').at(1);
    logoutLink.simulate('click');
    expect(logoutRequest).toHaveBeenCalled();
  });

  it('test handle key down', () => {
    const login = {
      isSignedIn: true
    };

    const history = { 'length': 3, 'action': 'PUSH', push: jest.fn(), 'location': { 'pathname': '/search', 'search': '', 'hash': '', 'key': 's7js0s' } };
    const searchVideo = jest.fn();
    const header = <Router>
      <Header
        login={login}
        searchVideo={searchVideo}
        history={history}
        {...props} />
    </Router>;

    const wrapper = mount(header);
    const searchInput = wrapper.find('input');
    searchInput.simulate('keyDown', { keyCode: 13 });
    expect(searchVideo).toHaveBeenCalled();
  });

  it('should render dropdown when signed in', () => {
    const login = {
      isSignedIn: true
    };
    const loginRequest = jest.fn();
    const wrapper = mount(<Router>
      <Header login={login} loginRequest={loginRequest} {...props} />
    </Router>);

    const btnLogin = wrapper.find('.btn-login').at(1);
    const dropdown = wrapper.find('.dropdown');
    expect(dropdown.exists()).toBe(true);
    expect(btnLogin.exists()).toBe(false);
  });

  it('should render btnLogin when not signed in', () => {
    const login = {
      isSignedIn: false
    };
    const loginRequest = jest.fn();
    const wrapper = mount(<Router>
      <Header login={login} loginRequest={loginRequest} {...props} />
    </Router>);

    const btnLogin = wrapper.find('.btn-login').at(1);
    const dropdown = wrapper.find('.dropdown');
    expect(btnLogin.exists()).toBe(true);
    expect(dropdown.exists()).toBe(false);
  });

  it('should render correctly', () => {
    const login = {
      isSignedIn: false
    };
    const loginRequest = jest.fn();
    const tree = renderer
      .create(<Router><Header login={login} loginRequest={loginRequest} {...props} /></Router>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});