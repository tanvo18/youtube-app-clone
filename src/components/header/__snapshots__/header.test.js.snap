// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test header should render correctly 1`] = `
<div
  className="container-fluid sc-bxivhb jKKnLa"
>
  <h1
    className="logo sc-bdVaJa ihUdBI"
  >
    <a
      href="/"
    >
      <img
        alt=""
        src="youtube.png"
      />
    </a>
  </h1>
  <div
    className="search-wrap"
  >
    <input
      className="sc-bwzfXH kkcLHG"
      height="2.375rem"
      onKeyDown={[Function]}
      placeholder="Search"
      type="text"
      width="100%"
    />
    <div
      className="input-group-append"
    >
      <a
        href="/search"
        onClick={[Function]}
      >
        <button
          className="btn sc-htpNat RZbAn"
          fontSize="16"
          onClick={[Function]}
        >
          <i
            className="fa fa-search"
          />
        </button>
      </a>
    </div>
  </div>
  <div
    className="login-btn-wrap"
  >
    <div>
      <button
        className="btn btn-login sc-htpNat kJLzS"
        fontSize="16"
        onClick={[Function]}
      >
        SIGN IN
      </button>
    </div>
  </div>
</div>
`;
